### PR TITLE
Support o1

### PR DESCRIPTION
**Pull request by PopDaph**
## Description

Review with hidden whitespaces advised: https://github.com/dust-tt/dust/pull/7506/files?diff=split&w=1

### Core
Adds support for O1 in Core, with some hacks since it does not support system messages nor streaming. 
I only support it on OpenAI not azureOpenAI, but we could of course easily do it if we wanted. 
I made a few tests with a convo with both O1 and an agent with websearch and it seems to work properly.
Will make more tests on prod before checking if we need to add more hacks (since it does not support tools).

### On front
- Add a global agent `@openai-o1`, behind a flag. 
- We cannot select this model on the Assistant Builder. 
- We can select this model from a Dust app.

### Fake streaming
I'm fake streaming from Core. View XP here: https://www.loom.com/share/a32028526f124b6c97076142fefaf621
=> Actually I added a commit to stream a bit faster.

## Risk

Can be rolled back.

## Deploy Plan

Deploy Core. 
Deploy front.

### CODE REVIEW

File: front/lib/api/assistant/global_agents.ts [hunk 1]
```diff
@@ -252,6 +253,43 @@ function _getGPT4GlobalAgent({
...
+  return {
+    id: -1,
+    sId: GLOBAL_AGENTS_SID.O1,
+    version: 0,
+    versionCreatedAt: null,
+    versionAuthorId: null,
+    name: "openai-o1",
+    description: O1_MODEL_CONFIG.description,
+    instructions: null,
+    pictureUrl: "https://dust.tt/static/systemavatar/gpt4_avatar_full.png", // @todo: change to O1 avatar
```
Comment (PopDaph) [inline front/lib/api/assistant/global_agents.ts:277]: Will see that with Ed, but not blocking to ship since feature flag anyway!

**Comment by PopDaph** [comment]
I wanted to release for us firs to have it tested by everyone.

Next step I will focus on defining what we want to do for this: 

- **Explaining limitations to users:** We know there are significant limitations on the API  side at the moment (no streaming, no streaming of chain of thought, etc…) so we’ll need to have appropriate caveats
- **Explaining limitations on usage: we have a capped # of requests per minute, and we’ll need to give users the visibility on errors that result from that in a way that’s understandable**

File: core/src/providers/openai.rs [hunk 2]
```diff
@@ -2008,72 +2012,98 @@ impl LLM for OpenAILLM {
...
+                        .map(|c| c.iter().collect::<String>())
+                    {
+                        let _ = sender.send(json!({
+                            "type": "tokens",
+                            "content": {
+                                "text": chunk,
+                            },
+                        }));
+                        // Add a small delay to simulate real-time streaming.
+                        tokio::time::sleep(std::time::Duration::from_millis(50)).await;
```
Comment (fontanierh) [inline core/src/providers/openai.rs]: as per IRL, a tiny bit faster might be nice

File: types/src/front/lib/assistant.ts [hunk 3]
```diff
@@ -87,6 +87,7 @@ export const GPT_4_TURBO_MODEL_ID = "gpt-4-turbo" as const;
 export const GPT_4O_LEGACY_MODEL_ID = "gpt-4o" as const;
 export const GPT_4O_MODEL_ID = "gpt-4o-2024-08-06" as const;
 export const GPT_4O_MINI_MODEL_ID = "gpt-4o-mini" as const;
+export const O1_MODEL_ID = "o1-preview" as const;
```
Comment (fontanierh) [inline types/src/front/lib/assistant.ts:90]: nit: should we call this `O1_PREVIEW_MODEL_ID` instead ? And maybe also support `O1_MINI_MODEL_ID` ?

**Review by fontanierh — APPROVED** [review]

File: core/src/providers/openai.rs (hunk 2 above)
Comment (PopDaph) [reply core/src/providers/openai.rs]: Done!

File: types/src/front/lib/assistant.ts (hunk 3 above)
Comment (PopDaph) [reply types/src/front/lib/assistant.ts]: Done for the rename, and I also added a global agent for mini, with a different feature flag so we can activate one or the other or both

File: types/src/front/lib/assistant.ts [hunk 4]
```diff
@@ -244,6 +248,36 @@ export const GPT_4O_MINI_MODEL_CONFIG: ModelConfigurationType = {
   toolUseMetaPrompt: LEGACY_OPEN_AI_TOOL_USE_META_PROMPT,
   supportsVision: true,
 };
+export const O1_PREVIEW_MODEL_CONFIG: ModelConfigurationType = {
+  providerId: "openai",
+  modelId: O1_PREVIEW_MODEL_ID,
+  displayName: "Open AI O1 Preview",
```
Comment (spolu) [inline types/src/front/lib/assistant.ts:254]: No space in OpenAI :)

Comment (spolu) [reply types/src/front/lib/assistant.ts:254]: Also we generally don't put it in the displayName since the logo is next to it?

Comment (PopDaph) [reply types/src/front/lib/assistant.ts:254]: https://github.com/dust-tt/dust/pull/7512

**Pull request merged**